### PR TITLE
Polyline functions

### PR DESCRIPTION
**Pull request by mattiskoh**
### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
 ### Added
+* Added `divide_polyline`, `divide_polyline_by_length`, `Polyline.split_at_corners` and `.Polyline.tangent_at_point_on_polyline`.
```
Comment (beverlylytle) [inline CHANGELOG.md]:
```suggestion
* Added `divide_polyline`, `divide_polyline_by_length`, `Polyline.split_at_corners` and `Polyline.tangent_at_point_on_polyline`.
```

File: src/compas/geometry/primitives/polyline.py [hunk 2]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
             self.points[index].y = point[1]
             self.points[index].z = point[2]
 
+    def split_at_corners(self, angle_threshold):
+        """Splits a polyline at corners larger than the given angle_threshold
 
+        Parameters:
+        -----------
+        angle_threshold: radians
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        angle_threshold : float
            In radians.
```

File: src/compas/geometry/primitives/polyline.py [hunk 3]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+
+        Returns
+        -------
+        list of polylines
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        list of :class:`compas.geometry.Polyline`
```

File: src/compas/geometry/primitives/polyline.py [hunk 4]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+                split_polylines.append(Polyline(looped_pts))
+
+        return split_polylines
+
+    def tangent_at_point_on_polyline(self, point):
+        """Calculates the tangent vector of a point on a polyline
+
+        Parameters:
+        -----------
+        point: :class:`compas.geometry.Primitives.Point`
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        point: :class:`compas.geometry.Point`
```

File: src/compas/geometry/primitives/polyline.py [hunk 5]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+
+        Returns
+        -------
+        Vector
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        :class:`compas.geometry.Vector`
```

File: src/compas/geometry/primitives/polyline.py [hunk 6]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+        """Divide a polyline in equal segments
+
+        Parameters:
+        -----------
+        num_segments: int
+
+        Returns
+        -------
+        list
+            list of points :class:`compas.geometry.Primitives.Point`
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
            list of :class:`compas.geometry.Point`
```

File: src/compas/geometry/primitives/polyline.py [hunk 7]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+        for line in self.lines:
+            if is_point_on_line(point, line):
+                return line.direction
+
+    def divide_polyline(self, num_segments):
+        """Divide a polyline in equal segments
+
+        Parameters:
+        -----------
+        num_segments: int
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        num_segments : int
```

File: src/compas/geometry/primitives/polyline.py [hunk 8]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+            division_pts.append(new_polyline.points[-1])
+
+        return division_pts
+
+    def divide_polyline_by_length(self, length):
+        """Splits a polyline in segments of a given length
+
+        Parameters:
+        -----------
+        length: float
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        length : float
```

File: src/compas/geometry/primitives/polyline.py [hunk 9]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+
+        Returns
+        -------
+        list
+            list of points :class:`compas.geometry.Primitives.Point`
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
            list of :class:`compas.geometry.Point`
```

File: src/compas/geometry/primitives/polyline.py [hunk 10]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+        else:
+            corner_ids = [0] + corner_ids + [len(points)]
+
+        for id1, id2 in pairwise(corner_ids):
+            if id1 < id2:
+                split_polylines.append(Polyline(points[id1:id2+1]))
+            else:
+                looped_pts = [points[i] for i in range(id1, len(points))] + points[1:id2+1]
+                split_polylines.append(Polyline(looped_pts))
+
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py:301]: If this is invoked on a closed Polyline that approximates, say, a circle with very many line segments and a sufficiently large `angle_threshold`, then this would return an empty list, and I think it makes sense to return a list containing only a copy of the original Polyline.
```suggestion
if self.is_closed() and not corner_ids:
    return [Polyline(self.points)]
    
```

File: src/compas/geometry/primitives/polyline.py [hunk 11]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+        if self.is_closed() and len(corner_ids) > 0:
+            if corner_ids[-1] != len(points):
+                corner_ids = [corner_ids[-1]] + corner_ids
+        else:
+            corner_ids = [0] + corner_ids + [len(points)]
+
+        for id1, id2 in pairwise(corner_ids):
+            if id1 < id2:
+                split_polylines.append(Polyline(points[id1:id2+1]))
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py:297]: Do you really want that the new `Polyline`s use the same `Point`s, or should the points be copied?

File: src/compas/geometry/primitives/polyline.py [hunk 12]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+                    remaining_pts = new_polyline.points[i_ln+2:]
+                    new_polyline = Polyline([new_pt, line.end] + remaining_pts)
+                    break
+            if len(division_pts) == num_pts+1:
+                break
+
+        if len(division_pts) != num_pts+1:
+            division_pts.append(new_polyline.points[-1])
+
+        return division_pts
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]: This code is just the same as below, right?  so to avoid duplication, just call the function below.
```suggestion
        return self.divide_polyline_by_length(segement_length)
```

File: src/compas/geometry/primitives/polyline.py [hunk 13]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+        -----------
+        point: :class:`compas.geometry.Primitives.Point`
+
+        Returns
+        -------
+        Vector
+        """
+        for line in self.lines:
+            if is_point_on_line(point, line):
+                return line.direction
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py:320]:
```suggestion
                return line.direction
            raise Exception('{} not found!'.format(point))
```

File: src/compas/geometry/primitives/polyline.py [hunk 14]
```diff
@@ -258,11 +259,134 @@ def transform(self, T):
...
+        Parameters:
+        -----------
+        length: float
+
+        Returns
+        -------
+        list
+            list of points :class:`compas.geometry.Primitives.Point`
+        """
+        num_pts = int(round(self.length/length))
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]: This line I find a little confusing.  First, I think `round` returns an `int` when the precision is not specified.  Secondly, `round` behaves scientifically rather than computer-sciencely in that `round(4.5) == 4` while `round(5.5) == 6`.  So in `divide_polyline` I don't think you always get exactly the number of segments that you specify, but maybe that's ok.  Would `int(self.length / length)` make sense here (which always rounds down)?

**Review by beverlylytle — CHANGES_REQUESTED** [review]
Looking pretty good.  I know it looks like a lot of comments, but mostly they are trivial things to do with docstring style. 

I think it would be a really good thing to add some tests for these new functions.  In particular, testing around some edge cases. For example `divide_polyline` on polylines with only two points or with many, on closed or not closed polylines, with few or many `num_segments`.  And similarly for `divide_polyline_by_length` with a length that divides (or not) the polyline's length, or exceeds it. Etc.

File: src/compas/geometry/primitives/polyline.py (hunk 11 above)
Comment (mattiskoh) [reply src/compas/geometry/primitives/polyline.py:297]: I have no strong opinion about this. I definitely understand the reason to copy these points to avoid any mishaps later on in the function, but as it seems to work I would just leave it as is. 

Performance-wise I thought this would be the most efficient ( but I'm definitely no expert 😂 )

If you feel that it would improve the code by copying feel free to add.

File: src/compas/geometry/primitives/polyline.py (hunk 12 above)
Comment (mattiskoh) [reply src/compas/geometry/primitives/polyline.py]: Correct, I will add the edge case for the Polyline existing of only 2pts (a Line..) to `self.divide_polyline_by_length` and simplify `self.divide_polyline` as above.

File: src/compas/geometry/primitives/polyline.py (hunk 14 above)
Comment (mattiskoh) [reply src/compas/geometry/primitives/polyline.py]: Yes, you're right. I will add the change

File: src/compas/geometry/primitives/polyline.py (hunk 11 above)
Comment (beverlylytle) [reply src/compas/geometry/primitives/polyline.py:297]: I am ambivalent about this, weighing performance against hard to spot bugs. So, I'd say leave it as it is until someone complains ;)

File: src/compas/geometry/primitives/polyline.py [hunk 15]
```diff
@@ -258,11 +259,123 @@ def transform(self, T):
...
+        point: :class:`compas.geometry.Point`
+
+        Returns
+        -------
+        :class:`compas.geometry.Vector`
+        """
+        for line in self.lines:
+            if is_point_on_line(point, line):
+                return line.direction
+            raise Exception('{} not found!'.format(point))
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]: I think this needs to have one less level indentation.  Sorry!
```suggestion
        raise Exception('{} not found!'.format(point))
```

File: tests/compas/geometry/test_polyline.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+
+@pytest.mark.parametrize('coords,expected', [
+    ([[0.0, 0.0, 0.0], [100.0, 0.0, 0.0]], [[0.0, 0.0, 0.0], [80.0, 0.0, 0.0], [100.0, 0.0, 0.0]]),
+])
+def test_polyline_divide_length_strict2(coords, expected):
+    assert expected == Polyline(coords).divide_polyline_by_length(80, False)
+
+
+@pytest.mark.parametrize('coords, input, expected', [
```
Comment (beverlylytle) [inline tests/compas/geometry/test_polyline.py]:
```suggestion
@pytest.mark.parametrize('coords,input,expected', [
```

File: tests/compas/geometry/test_polyline.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    ([[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [1.0, 1.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 0.0]], math.pi/2, [Polyline([[0.0, 0.0, 0.0], [1, 0.0, 0.0]]),
+                                                                                                        Polyline([[1, 0.0, 0.0], [1, 1, 0.0]]), Polyline([[1, 1, 0.0], [0.0, 1, 0.0]]), Polyline([[0.0, 1, 0.0], [0.0, 0.0, 0.0]])]),
+])
+def test_polyline_split_at_corners(coords, input, expected):
+    assert expected == Polyline(coords).split_at_corners(input)
+
+
+@pytest.mark.parametrize('coords, input, expected', [
```
Comment (beverlylytle) [inline tests/compas/geometry/test_polyline.py]:
```suggestion
@pytest.mark.parametrize('coords,input,expected', [
```

File: src/compas/geometry/primitives/polyline.py [hunk 18]
```diff
@@ -258,11 +259,131 @@ def transform(self, T):
...
+        return self.divide_polyline_by_length(segment_length, False)
+
+    def divide_polyline_by_length(self, length, strict=True):
+        """Splits a polyline in segments of a given length
+
+        Parameters:
+        -----------
+        length: float
+
+        strit: boolean
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
        strict : bool
```

File: src/compas/geometry/primitives/polyline.py [hunk 19]
```diff
@@ -258,11 +259,131 @@ def transform(self, T):
...
+            If set to False, the remainder segment will be added even if it is smaller than the desired length
```
Comment (beverlylytle) [inline src/compas/geometry/primitives/polyline.py]:
```suggestion
            If set to ``False``, the remainder segment will be added even if it is smaller than the desired length
```

**Review by beverlylytle — APPROVED** [review]
Lookin good!  let's merge this!
